### PR TITLE
Fix issue 18890 - extern(C++) mangles all destructors the same

### PR DESCRIPTION
**Pull request by TurkeyMan**
With `extern(C++)`, classes with multiple destructors were all mangling the same.
This patch only mangles the full-destructor as extern(C++), which is the behaviour C++ expects.

### CODE REVIEW

File: src/dmd/dsymbol.d
Comment (jacob-carlborg) [inline src/dmd/dsymbol.d]: @ibuclaw does this need to be added to C++ header?

Comment (TurkeyMan) [reply src/dmd/dsymbol.d]: I don't think so...?

Comment (jacob-carlborg) [inline src/dmd/dsymbol.d]: I would be nice with a description of what a primary destructor is.

Comment (TurkeyMan) [reply src/dmd/dsymbol.d]: I could... but this is the only PR that has passed CI after many hours.
starting over feels... disappointing :/

Comment (TurkeyMan) [reply src/dmd/dsymbol.d]: It is the outer/main destructor. Aggregates gather a list of destructors, but only one of them is the most inclusive destructor, ie `__xdtor`.

**Comment by jacob-carlborg** [comment]
I think this looks fine, but I'm a bit unsure about the C++ header file.

File: src/dmd/dsymbol.d
Comment (TurkeyMan) [reply src/dmd/dsymbol.d]: I can remove this function completely and roll the logic into a simpler one-lineer at each of the call-sites.

Comment (jacob-carlborg) [reply src/dmd/dsymbol.d]: What about adding a function in `cppmangle`? That can be imported to `cppmanglewin`. Or we wait for Iain.

**Comment by TurkeyMan** [comment]
I moved it to `cppmangle`.

File: src/dmd/cppmangle.d
Comment (jacob-carlborg) [inline src/dmd/cppmangle.d]: Ideally it should be `package` but I'll approve anyway.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by TurkeyMan** [comment]
Does this please everyone now?

File: src/dmd/cppmangle.d
Comment (JinShil) [reply src/dmd/cppmangle.d]: Yeah, make it `package` or document with all the appropriate `returns` and `params` DDoc sections.  And throw a newline
somewhere in the summary so it's not so long.

**Review by JinShil — DISMISSED** [review]

**Comment by TurkeyMan** [comment]
Right... just like the 2 functions immediately above it! ;)
Done. But now, this little fastidious request reboots the CI process, and wastes hours of my time. I just proliferated the pattern from the function immediately above.

**Comment by TurkeyMan** [comment]
Well choose your battles, and don't waste my fucking time, or we'll struggle to retain new contributors.
I'm not enjoying this process, and I'm not motivated to continue doing so. If you assert that the process is annoying, then I'll gladly go away.
The process is rather painful enough thanks to the CI situation, don't aggravate it with unnecessary CI cycles.

It's taken me 3 days (so far, still ongoing!) to merge 2 or 3 hours work. That is REALLY BAD value-for-time-spent ratio.

I was effectively blocked on the weekend with 6 or so outstanding PR's that all conflicted with eachother.
I can't just walk away and let it take care of itself, because every merge I have to rebase everything.

Ideally `package` could have been integrated into a prior rebase opportunistically, rather than adding an extra one to the end >_<

**Comment by TurkeyMan** [comment]
Good thing my mates `inglebrook` and `win-farm-2`are on it! :P

**Comment by JinShil** [comment]
> If you assert that the process is annoying...

I wasn't asserting that the process is annoying (though that is also true),  I was asserting that the inconsistencies in the code are annoying; we're incrementally submitting [refactoring PRs](https://github.com/dlang/dmd/pulls?q=is%3Apr+label%3ARefactoring+is%3Aclosed) in an attempt to deal with that.

> It's taken me 3 days (so far, still ongoing!)

3 days! That's all!  You're doing great!  I have had PRs that sat around for months without attention, and currently have PRs going on their 3rd week.  There are other contributors that currently have months-old PRs waiting for review, and yet others that are years old.  And, as you know all too well, these PRs constantly require babysitting to make sure they are rebased when they get clobbered by another PR, or when the CIs start throwing in spurious errors just for added fun.

It's a problem all of us are frustrated with, not just you, and quite honestly, as bad as it is now, it's better than it was 6 months ago, believe it or not.  The PR queue used to be at over 190, until myself and a few others joined, and that was while keeping pace with a larger than normal increase in contributions.  See the [2nd graph here around Dec/Jan](https://auto-tester.puremagic.com/chart.ghtml?projectid=1).  Before then you'd be lucky to have anyone look at your PR in a week, let a lone 24 hours.

**Review by dnadlinger — APPROVED** [review]

**Comment by TurkeyMan** [comment]
> I have had PRs that sat around for months without attention, and currently have PRs going on their 3rd week.

I understand that is the experience people have. It's also a little bit circumstantial; in my case now, I don't believe anything I'm doing is controversial, in theory or in implementation, so that should ideally lead to short turn-around time.
It also has this 'vision' tag which seems to help.

> Before then you'd be lucky to have anyone look at your PR in a week, let a lone 24 hours.

I'm not trying to suggest that I should require special treatment, I'm just trying to point out that the process is difficult to accept, and we have a known problem with contributor motivation in general.
I can safely say that in a very practical sense, as soon as I'm not overtly motivated to fix my things for my own use, I will not persist as a contributor.
That's not a good place to be, and it's not a good experience for potential contributors to have up-front.

If anything, D is held back by number of contributors more than anything. We need to get much better at this.

Interestingly, in my experience so far, review time is not a problem, it is CI time mostly. Reviewers have generally been very prompt, so we seem to be doing well there. What sucks is the last-minute request for trivial change. You could have clicked the button, but instead clicked the red X... 3 hours were eliminated from my schedule at that moment. We only live once, and not for very long.
If reviews have happened, and CI is finished, I think you need to be REALLY careful about asking for trivial changes like that. I clearly made best-effort to match surrounding code style.

Anyway, it's fine, I did the thing, we're done here... it's almost 3 hours later.

**Comment by JinShil** [comment]
> If anything, D is held back by number of contributors more than anything

Actually, I argue it's held up by the number of senior reviewers.  I could easily fill the PR queue myself with contributions, but that will just take resources away from other PRs that I believe are more important.

There are only a few people on our team that can certify certain PRs due to the domain expertise required, and unfortunately, they are very part-time reviewers.  If we could get more senior reviewers on the team (or get our existing reviewers to volunteer more of their time), the PR process will greatly improve, and that will elicit many more contributions and contributors.  I'm hoping, once @WalterBright finishes converting the backend to D, he'll be willing to spend more of his time reviewing PRs from others; we desperately need someone with his skills tending to the PR queue regularly.

**Comment by wilzbach** [comment]
> this little fastidious request reboots the CI process, and wastes hours of my time. I just proliferated the pattern from the function immediately above.

Could you please explain why this is taking up hours of your time?
Running the entire testsuite takes about 10 seconds on my local machine and if you haven't done major changes that would affect Windows (i.e. such stylistic quick edits), then it shouldn't take you more than one minute with the git commit and git push, no?
I hope you aren't watching the CIs, are you?

**Comment by TurkeyMan** [comment]
Because I had a queue of 5 PR's that all depended/conflicted on eachother. Each time one got the magic auto-merge tag, I had to stop whatever I was doing, and rebase everything to keep the process moving.

I totally watch the CI's, because once any one of those patches merged, I had to do work to fix the rest.

**Comment by wilzbach** [comment]
Ah I see, but I'm not sure there's anything we can do to help this. Some machines on the auto-tester are just pretty busy :/
(yeah we could add more machines, but this infrastructure isn't under our control).
If you other suggestions that would help you, don't hesitate to let me/us know.

> merged, I had to do work to fix the rest.

Don't you get a email notification?

**Comment by TurkeyMan** [comment]
I'm just suggesting that when there are lots of machines that can build in <10 minutes, that you don't ALSO want machines that take >50minutes.
If one of those slow builds are pending at the time you add auto-merge (there always is, that's the nature of the one that takes a long time... it takes a long time), then the merge doesn't complete for an extra hour than it needs to.
I was just trying to work through the things, and I'd rather 5 merges take 5*10 minutes, than 5*50 minutes. One feels productive, the other feels agnoising.

**Pull request merged**
